src/template/wasm2c.declarations.c

```diff
@@ -20,6 +20,40 @@
 #define MEM_ADDR(mem, addr, n) &(mem)->data[addr]
 #endif
 
+#ifndef WASM_RT_USE_SEGUE
+// Memory functions can use the segue optimization if allowed. The segue
+// optimization uses x86 segments to point to a linear memory. We use this
+// optimization when:
+//
+// (1) Segue is allowed using WASM_RT_ALLOW_SEGUE
+// (2) on x86_64 without WABT_BIG_ENDIAN enabled
+// (3) the Wasm module uses a single unshared imported or exported memory
+// (4) the compiler supports: intrinsics for (rd|wr)gsbase, "address namespaces"
+//     for accessing pointers, and supports memcpy on pointers with custom
+//     "address namespaces". GCC does not support the memcpy requirement, so
+//     this leaves only clang for now.
+// (5) The OS doesn't replace the segment register on context switch which
+//     eliminates windows for now
+#if WASM_RT_ALLOW_SEGUE && !WABT_BIG_ENDIAN &&               \
+    (defined(__x86_64__) || defined(_M_X64)) &&              \
+    WASM_RT_MODULE_IS_SINGLE_UNSHARED_MEMORY && __clang__ && \
+    __has_builtin(__builtin_ia32_wrgsbase64) && !defined(_WIN32)
+#define WASM_RT_USE_SEGUE 1
+#else
+#define WASM_RT_USE_SEGUE 0
+#endif
+#endif
+
+#if WASM_RT_USE_SEGUE
+// POSIX uses FS for TLS, GS is free
+#define WASM_RT_SEGUE_READ_BASE() __builtin_ia32_rdgsbase64()
+#define WASM_RT_SEGUE_WRITE_BASE(base) \
+  __builtin_ia32_wrgsbase64((uintptr_t)base)
+#define MEM_ADDR_MEMOP(mem, addr, n) ((uint8_t __seg_gs*)(uintptr_t)addr)
+#else
+#define MEM_ADDR_MEMOP(mem, addr, n) MEM_ADDR(mem, addr, n)
```

Comment:
Do we need to introduce another macro name here?

Can not just redefine `MEM_ADDR` here and avoid changing the LOAD/STORE macros below?

Reply:
Unfortunately, this isn't possible as MEM_ADDR is used in a lot of locations --- shared memory ops, simd memory ops, wasm data section initialization etc.  which we don't want to modify. Hence we need a separate macro

Reply:
I see.. that makese sense.  But it wasn't obvious to me.  Perhaps we could add a comment here or maybe we can think of a more descriptive name?  `LOAD_STORE_ADDR`?     Not a big deal though.